server/src/main/java/io/deephaven/server/console/python/DebuggingInitializer.java

```diff
@@ -0,0 +1,45 @@
+package io.deephaven.server.console.python;
+
+import io.deephaven.configuration.Configuration;
+import io.deephaven.util.thread.ThreadInitializationFactory;
+import org.jpy.PyLib;
+import org.jpy.PyModule;
+import org.jpy.PyObject;
+
+import java.io.Closeable;
+
+/**
+ * If python is configured as the script language for this server, ensures that threads which may invoke python code
+ * will be able to be debugged. If python is disabled, this does nothing.
+ */
+public class DebuggingInitializer implements ThreadInitializationFactory {
+    @Override
+    public Runnable createInitializer(Runnable runnable) {
+        if (!"python".equals(Configuration.getInstance().getStringWithDefault("deephaven.console.type", null))) {
+            // python not enabled, don't accidentally start it
+            return runnable;
+        }
+        DeephavenModule py_deephaven = (DeephavenModule) PyModule.importModule("deephaven")
+                .createProxy(PyLib.CallableKind.FUNCTION, DeephavenModule.class);
+
+
+        return () -> {
+            // First call in to create a custom function that has the same name as the Java thread (plus a prefix)
+            PyObject runnableResult = py_deephaven.create_thread_entry(Thread.currentThread().getName());
+            // runnable.run();
+            // Invoke that function directly from Java, so that we have only this one initial frame
+            runnableResult.call("__call__", runnable);
```

Comment:
This technique of ensuring the java frame includes python in the Runnable stack _does_ rely on the fact that python drops the GIL when calling into java. https://github.com/jpy-consortium/jpy/pull/48 nothing actionable, but good to note it wouldn't work otherwise.

Reply:
True, but the fact that DHaaL works at all relies on that fact as well, the entire library would deadlock on the gil and ugp if we didn't make this assumption.